File: gitlab_downloader/reporting.py
from __future__ import annotations

import json
import logging
from datetime import datetime, timezone
from pathlib import Path

from .models import CloneResult, GitlabConfig

logger = logging.getLogger("gitlab_downloader")


def print_summary(results: list[CloneResult]) -> bool:
    success = sum(1 for item in results if item.status == "success")
    skipped = sum(1 for item in results if item.status == "skipped")
    failed = sum(1 for item in results if item.status == "failed")
    updated = sum(1 for item in results if item.status == "updated")

    logger.info(
        "Summary: success=%s updated=%s skipped=%s failed=%s",
        success,
        updated,
        skipped,
        failed,
    )

    if failed:
        logger.error("Failed repositories:")
        for item in results:
            if item.status == "failed":
                logger.error("- %s: %s", item.name, item.message)

    return failed > 0


def print_dry_run(projects: list[dict], config: GitlabConfig, build_clone_target) -> None:
    logger.info("Dry-run mode enabled. Projects to process: %s", len(projects))
    logger.info("%-8s %-30s %-30s %-45s %s", "ID", "NAME", "GROUP_PATH", "URL", "TARGET")

    for project in projects:
        repo_name, target_path = build_clone_target(project, config)
        url = str(project.get("http_url_to_repo", ""))[:45]
        logger.info(
            "%-8s %-30s %-30s %-45s %s",
            str(project.get("id", "")),
            repo_name[:30],
            str(project.get("group_path", ""))[:30],
            url,
            target_path,
        )


def write_json_report(
    path: str, config: GitlabConfig, projects_count: int, results: list[CloneResult]
) -> None:
    payload = {
        "generated_at": datetime.now(timezone.utc).isoformat(),
        "group": config.group,
        "projects_count": projects_count,
        "summary": {
            "success": sum(1 for item in results if item.status == "success"),
            "updated": sum(1 for item in results if item.status == "updated"),
            "skipped": sum(1 for item in results if item.status == "skipped"),
            "failed": sum(1 for item in results if item.status == "failed"),
        },
        "results": [item.__dict__ for item in results],
    }

    output = Path(path)
    output.parent.mkdir(parents=True, exist_ok=True)
    output.write_text(json.dumps(payload, ensure_ascii=True, indent=2), encoding="utf-8")
    logger.info("JSON report written to %s", output)

File: lib/reporting.js
import { writeFile, mkdir } from 'node:fs/promises';
import { dirname } from 'node:path';

/**
 * Count results by status.
 *
 * @param {Array<{name: string, status: string, message: string}>} results
 * @returns {{success: number, updated: number, skipped: number, failed: number}}
 */
function countByStatus(results) {
  const counts = { success: 0, updated: 0, skipped: 0, failed: 0 };
  for (const item of results) {
    if (item.status in counts) {
      counts[item.status]++;
    }
  }
  return counts;
}

/**
 * Print a summary of clone/pull results.
 *
 * Returns true if there were any failures, false otherwise.
 *
 * @param {Array<{name: string, status: string, message: string}>} results
 * @param {object} [options]
 * @param {(msg: string) => void} [options.log] - Logging function (default: console.log)
 * @param {(msg: string) => void} [options.error] - Error logging function (default: console.error)
 * @returns {boolean} true if any results have status "failed"
 */
export function printSummary(results, options = {}) {
  const log = options.log || console.log;
  const error = options.error || console.error;
  const { success, updated, skipped, failed } = countByStatus(results);

  log(`Summary: success=${success} updated=${updated} skipped=${skipped} failed=${failed}`);

  if (failed > 0) {
    error('Failed repositories:');
    for (const item of results) {
      if (item.status === 'failed') {
        error(`- ${item.name}: ${item.message}`);
      }
    }
  }

  return failed > 0;
}

/**
 * Print a dry-run preview of planned operations in tabular format.
 *
 * @param {Array<object>} projects - GitLab project objects
 * @param {object} config - GitlabConfig
 * @param {(project: object, config: object) => {repoName: string, targetPath: string}} buildCloneTarget - Function to compute clone target
 * @param {object} [options]
 * @param {(msg: string) => void} [options.log] - Logging function (default: console.log)
 */
export function printDryRun(projects, config, buildCloneTarget, options = {}) {
  const log = options.log || console.log;

  log(`Dry-run mode enabled. Projects to process: ${projects.length}`);
  log(
    'ID'.padEnd(8) +
    'NAME'.padEnd(30) +
    'GROUP_PATH'.padEnd(30) +
    'URL'.padEnd(45) +
    'TARGET'
  );

  for (const project of projects) {
    const { repoName, targetPath } = buildCloneTarget(project, config);
    const url = (project.http_url_to_repo || '').slice(0, 45);

    log(
      String(project.id || '').padEnd(8) +
      repoName.slice(0, 30).padEnd(30) +
      (project.group_path || '').slice(0, 30).padEnd(30) +
      url.padEnd(45) +
      targetPath
    );
  }
}

/**
 * Write a JSON report file with clone/pull results.
 *
 * Creates parent directories if they do not exist.
 *
 * @param {string} path - Output file path
 * @param {object} config - GitlabConfig (needs .group)
 * @param {number} projectsCount - Total number of projects
 * @param {Array<{name: string, status: string, message: string}>} results
 */
export async function writeJsonReport(path, config, projectsCount, results) {
  const payload = {
    generated_at: new Date().toISOString(),
    group: config.group || null,
    projects_count: projectsCount,
    summary: countByStatus(results),
    results: results.map((item) => ({ ...item })),
  };

  const dir = dirname(path);
  await mkdir(dir, { recursive: true });
  await writeFile(path, JSON.stringify(payload, null, 2), 'utf-8');
}
